SPEC.md

```diff
@@ -0,0 +1,121 @@
+Keys may contain dots, but may not start or end with them.
+Two dots in a row are disallowed (due to being nonsensical), but implementations do not need to explicitly disallow them.
+
+### Section Declaration
+A section consists of an opening bracket (`[`), the section name, and a closing bracket (`]`).
+The section name may either be a valid key, or empty.
+
+Until the next section, all keys will be automatically prepended the section name and a `.`, or with nothing (if the section name is the empty string).
+
+### Key-(Raw)Value Pairs
+A key-value and key-rawvalue pair look very similar.
+It's a key, an equals, and a value, with arbitrary amounts of spaces before and after the `=`.
+
+A value may contain any UTF-8 characters except vertical whitespace or the `#` sign (note: this is expanded upon in ini-compatibility).
+It may not start or end with whitespace, or start with a backtick.
+
+A rawvalue starts with a backtick and ends with a backtick, but may contain any UTF-8 characters besides those.
+If a backtick needs to be inserted into a raw value, a backtick may be escaped with a second one.
+
+Theoretically, this means that both values and raw values may contain arbitrary bytes, but the specification is exclusively intended to be used with UTF-8 data.
+Implementations are not required to validate it, though.
+
+### Comments
+A comment starts with a `#` (note: this is expanded upon in ini-compatibility) and ends at the next newline.
+A comment may start on its own line, at any point inside of a value, after a section declaration, or after a raw value.
+It is impossible to have comments inside of a raw value, or inside of a section declaration.
+
+## Ini-Compatibility
+For compatibility with INI, the `;` character may also be used for comments.
+In this case, anytime the `#` is a significant character in the core language, the `;` character should be as well.
+
+## Extensions
+All extensions are fully optional.
+Currently, multiple of them are consequences of keeping the grammar lean, and are documented, but only parts of their behaviors are desirable.
+
+### Flexspace
+In implementations with the "flexspace" extension, whitespace within statements is not significant.
+This means that you may have arbitrary whitespace (not only spaces, but also tabs, newlines, zero-width spaces, and so on) in the following locations:
+* Between the opening section bracket, the section name, and the closing section bracket.
+* Between the key in a key-(raw)value pair, the `=`, and the (raw)value.
+* Between a (raw) value and a comment.
+
+### Tabulation
+In implementations with the "tabulation" extension, whitespace before and after statements is not significant.
+This means you may have arbitrary whitespace in the following locations:
+* In front of a section declaration.
+* In front of a key-(raw)value pair.
+* In front of a comment.
+* Between a key-rawvalue pair and the next statement.
+* Between a section declaration and the next statement.
```

Comment:
Does this mean that `core` does not allow empty lines? On the other hand, the name suggests this should only allow for arbitrary *horizontal* whitespace.
The lines between `core`, `flexspace` and `tabulation` are a bit blurry here.

Reply:
You're right, this should be *horizontal* whitespace.
`core` allows empty lines, and in fact core tests have empty lines (this is a mild reminder that the test suite and ref. impl is still the source of truth, SPEC.md is here to help people interpret the two of them).